### PR TITLE
nimbo: init at 0.2.1

### PR DESCRIPTION
**Pull request by ahldr**
###### Motivation for this change

Nimbo is a dead-simple wrapper for the AWS CLI that allows you to run jobs on AWS with a single command. All you need is a tiny config file, and Nimbo handles all the instance, environment, data, and IAM management. It also provides many useful commands to make it easier than ever to work with AWS from your terminal.

Note that this is an unofficial nix-wrapping, I'm not a Nimbo developer or claim to speak for them---just a fan.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md). (hopefully 😅)

### CODE REVIEW

File: pkgs/development/python-modules/nimbo/default.nix [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  };
+
+  propagatedBuildInputs = with python3Packages; [ setuptools boto3 awscli requests click pyyaml pydantic ];
+
+  # nimbo tests require an AWS instance
+  doCheck = false;
+
+  meta = with lib; {
+    description = "Run machine learning jobs on AWS with a single command";
+    homepage = "https://test.pypi.org/project/nimbo";
```
Comment (fabaff) [inline pkgs/development/python-modules/nimbo/default.nix:25]: Use the project's homepage. This saves often a click ;-)

```suggestion
    homepage = "https://github.com/nimbo-sh/nimbo";
```

File: maintainers/maintainer-list.nix [hunk 2]
```diff
@@ -428,6 +428,12 @@
     githubId = 782180;
     name = "Alex Vorobiev";
   };
+  alex-eyre = {
```
Comment (fabaff) [inline maintainers/maintainer-list.nix:431]: Please move this change to a separate commit in this PR.

File: pkgs/development/python-modules/nimbo/default.nix [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  pname = "nimbo";
+  version = "0.2.2";
+
+  disabled = pythonOlder "3.6";
+
+  src = fetchFromGitHub {
+    owner = "nimbo-sh";
+    repo = "nimbo";
+    sha256 = "0qagib2607qvp5fl71325qpyrpvfhgmk1zix0nbws91n6i3fzc6i";
+    rev = "bfe1e691469011f0fd9c64856a47b07493556b4b"; # up-to-date 2021-04-28
```
Comment (fabaff) [inline pkgs/development/python-modules/nimbo/default.nix:15]: Request to tag the source: https://github.com/nimbo-sh/nimbo/issues/4

File: pkgs/development/python-modules/nimbo/default.nix [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+  };
+
+  propagatedBuildInputs = with python3Packages; [ setuptools boto3 awscli requests click pyyaml pydantic ];
+
+  # nimbo tests require an AWS instance
+  doCheck = false;
```
Comment (fabaff) [inline pkgs/development/python-modules/nimbo/default.nix:21]: Can we run `pythonImportsCheck`?

File: pkgs/development/python-modules/nimbo/default.nix [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+buildPythonApplication rec {
+  pname = "nimbo";
+  version = "0.2.2";
+
+  disabled = pythonOlder "3.6";
+
+  src = fetchFromGitHub {
+    owner = "nimbo-sh";
+    repo = "nimbo";
```
Comment (fabaff) [inline pkgs/development/python-modules/nimbo/default.nix:13]: Opinionated change...

```suggestion
    repo = pname;
```

File: pkgs/development/python-modules/nimbo/default.nix [hunk 6]
```diff
@@ -0,0 +1,29 @@
+{ lib, python3Packages, fetchFromGitHub, awscli }:
+
+with python3Packages;
+
+buildPythonApplication rec {
```
Comment (fabaff) [inline pkgs/development/python-modules/nimbo/default.nix:5]: This is a often used pattern for Python application (at least recently but there are others):

```bash
{ lib
, fetchFromGitHub
, python3
}:

python3.pkgs.buildPythonApplication rec {
  pname = "xyz";
  version = "5.0.0";
  disabled = python3.pythonOlder "3.6";
[...]
  propagatedBuildInputs = with python3.pkgs; [
    beautifulsoup4
  ];
```

File: pkgs/development/python-modules/nimbo/default.nix (hunk 3 above)
Comment (noreferences) [reply pkgs/development/python-modules/nimbo/default.nix:15]: tag "v0.2.1" is ready to be used here

File: pkgs/development/python-modules/nimbo/default.nix [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+  pname = "nimbo";
+  version = "0.2.2";
```
Comment (noreferences) [inline pkgs/development/python-modules/nimbo/default.nix:7]: Use version "0.2.1" for now

File: pkgs/development/python-modules/nimbo/default.nix [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+  propagatedBuildInputs = with python3Packages; [ setuptools boto3 awscli requests click pyyaml pydantic ];
+
+  # nimbo tests require an AWS instance
+  doCheck = false;
+
+  meta = with lib; {
+    description = "Run machine learning jobs on AWS with a single command";
+    homepage = "https://test.pypi.org/project/nimbo";
+    license = licenses.bsl11;
+    maintainers = with maintainers; [ alex-eyre ];
```
Comment (noreferences) [inline pkgs/development/python-modules/nimbo/default.nix:27]: I'm happy to add myself as a maintainer since I use NixOS as my daily driver.

**Comment by alex-eyre** [comment]
Thanks for the reviews guys, I'll implement these once I'm back in front of a computer

**Comment by noreferences** [comment]
I just tested this on my NixOS (with `useSandbox`), works fine. @alex-eyre you can add a tick to the list.

File: pkgs/development/python-modules/nimbo/default.nix [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+  pname = "nimbo";
+  version = "v0.2.1";
+
+  disabled = python3.pythonOlder "3.6";
+
+  src = fetchFromGitHub {
+    owner = "nimbo-sh";
+    repo = pname;
+    sha256 = "0qagib2607qvp5fl71325qpyrpvfhgmk1zix0nbws91n6i3fzc6i";
+    rev = version;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/nimbo/default.nix:13]:
```suggestion
  version = "0.2.1";

  disabled = python3.pythonOlder "3.6";

  src = fetchFromGitHub {
    owner = "nimbo-sh";
    repo = pname;
    sha256 = "0qagib2607qvp5fl71325qpyrpvfhgmk1zix0nbws91n6i3fzc6i";
    rev = "$v{version}";
```

File: pkgs/development/python-modules/nimbo/default.nix [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    awscli
+    requests
+    click
+    pyyaml
+    pydantic
+  ];
+
+  # nimbo tests require an AWS instance
+  doCheck = false;
+  pythonImportsCheck = [ "botocore" "click" "yaml" "requests" ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/nimbo/default.nix:28]:
```suggestion
  pythonImportsCheck = [ "nimbo" ];
```

File: pkgs/development/python-modules/nimbo/default.nix (hunk 8 above)
Comment (SuperSandro2000) [reply pkgs/development/python-modules/nimbo/default.nix:34]:
```suggestion
    maintainers = with maintainers; [ alex-eyre noreferences ];
```

File: pkgs/development/python-modules/nimbo/default.nix [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+  disabled = python3.pythonOlder "3.6";
+
+  src = fetchFromGitHub {
+    owner = "nimbo-sh";
+    repo = pname;
+    sha256 = "0qagib2607qvp5fl71325qpyrpvfhgmk1zix0nbws91n6i3fzc6i";
+    rev = version;
+  };
+
+  propagatedBuildInputs = with python3.pkgs; [
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/nimbo/default.nix:16]:
```suggestion
  propagatedBuildInputs = [
```

File: pkgs/development/python-modules/nimbo/default.nix [hunk 12]
```diff
@@ -0,0 +1,36 @@
+{ lib, fetchFromGitHub, python3, awscli }:
+
+python3.pkgs.buildPythonApplication rec {
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/nimbo/default.nix:3]:
```suggestion
{ lib, buildPythonApplication, pythonOlder, fetchFromGitHub, setuptools, boto3, awscli, requests, click, pyyaml, pydantic, awscli }:

buildPythonApplication rec {
```

File: pkgs/development/python-modules/nimbo/default.nix [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+  pname = "nimbo";
+  version = "v0.2.1";
+
+  disabled = python3.pythonOlder "3.6";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/nimbo/default.nix:7]:
```suggestion
  disabled = pythonOlder "3.6";
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/nimbo/default.nix (hunk 9 above)
Comment (alex-eyre) [reply pkgs/development/python-modules/nimbo/default.nix]: fixed in b6fca896de9f4b08d8b729c813be03758e8113df

File: pkgs/development/python-modules/nimbo/default.nix (hunk 12 above)
Comment (alex-eyre) [reply pkgs/development/python-modules/nimbo/default.nix]: This is a python application and is called from `all-packages.nix`, I don't know if I can use this style of arguments unless it's called from `python-packages.nix`?

**Comment by noreferences** [comment]
@alex-eyre you can now use Nimbo version `v0.2.4`

File: pkgs/development/python-modules/nimbo/default.nix [hunk 14]
```diff
@@ -0,0 +1,36 @@
+{ lib, python3, fetchFromGitHub, awscli }:
+with python3.pkgs;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/nimbo/default.nix:2]:
```suggestion

```

This is not allowed. Please list the packages and functions individually.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
please squash all package related commits so that we have one for the package and two for the maintainers.

File: pkgs/development/python-modules/nimbo/default.nix (hunk 12 above)
Comment (jonringer) [reply pkgs/development/python-modules/nimbo/default.nix]: IIRC, there's a preference to bring in packages from the related package scope. In this case `pkgs.callPackage`. So the existing inputs make sense.

Comment (jonringer) [reply pkgs/development/python-modules/nimbo/default.nix]: I stand corrected, this is in `python-modules`, yes, please import the individual python packages.

Comment (jonringer) [reply pkgs/development/python-modules/nimbo/default.nix]: > This is a python application and is called from `all-packages.nix`, I don't know if I can use this style of arguments unless it's called from `python-packages.nix`?

If you want this as a module and an application, do you mind adding the following to `pkgs/top-level/all-packages.nix`:
```
  <pkg-name> = with python3Packages; toPythonApplication <python-module>;
```
More information on this can be found [here](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#topythonapplication-function)

If you want it just as an application, then it shouldn't go into `python-modules`, but a related `pkgs/` directory.

File: pkgs/development/python-modules/nimbo/default.nix (hunk 14 above)
Comment (jonringer) [reply pkgs/development/python-modules/nimbo/default.nix]: For further explanation:

Using `pythonXPackages` isn't allowed from `python-modules` as this will mix interpreter versions for a given package. For example, if you were to do `python37Packages.<mypkg>`, but the expression pulled from `python3[8]Packages`. Python will be mis-importing several libraries, and this will likely cause issues at runtime.

Please refer to [Python dependency handling](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#handling-dependencies) on the best practices for handling dependencies within `python-modules`. Namely, you should be importing each python package individually, and not refer to another python package set (E.g. `python3Packages`).

If you're new to nixpkgs, you're free to view some of my videos on nixpkgs and python:
- [Python packaging using nixpkgs](https://www.youtube.com/watch?v=jXd-hkP4xnU)
- [Adding a package to nixpkgs unstable](https://www.youtube.com/watch?v=fvj8H5yUKu8)

**Comment by jonringer** [comment]
everything alright @alex-eyre ?
